src/sema/mod.rs

```diff
@@ -50,6 +50,14 @@ pub fn sema(file: &ResolvedFile, resolver: &mut FileResolver, ns: &mut ast::Name
         // Checks for unused variables
         check_unused_namespace_variables(ns);
         check_unused_events(ns);
+
+        // Substrate: Every contrat must have at least one public message
+        if ns.target.is_substrate() && !ns.contracts.iter().any(|c| c.has_public_functions(ns)) {
+            ns.diagnostics.push(Diagnostic::warning(
+                pt::Loc::CommandLine,
+                format!("{} yields no public messages", file.full_path.display()),
+            ))
+        }
```

Comment:
I see your point and I already tried to implement it more or less like that yesterday. The problem I faced is how would you detect that a contract, while it may not have any public methods, when it is used as a base contract it is fine. So actually we can not decide whether a contract is valid or not by looking at them individually, it would be wrong to `error` or even `warn` about a base contract with no public functions. So instead I came to the conclusion in `sema`, logically it would be correct to warn if there is no public message __at all__.

Or is there an easy and inexpensive way to check whether a contract is used as a base I missed out?

Reply:
As discussed in our daily, I implemented it now so that it warns the user that any contract without external messages should be declared `abstract`.

This got me thinking, could we actually `error` on that? Only because it is allowed in `solc` does not mean we ultimately must allow it too. If this would result into an error in `sema` we could spare us various redundant checks throughout `codegen` and metadata generation. @seanyoung what do you think?